src/System.Windows.Forms.Primitives/src/System/Windows/Forms/Internals/DpiHelper.DpiAwarenessScope.cs

```
@@ -0,0 +1,106 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.ComponentModel;
+using System.Runtime.InteropServices;
+using System.Windows.Forms.Primitives.Resources;
+
+namespace System.Windows.Forms
+{
+    internal static partial class DpiHelper
+    {
+        /// <summary>
+        ///  Class that help setting <see cref="DPI_AWARENESS_CONTEXT"/> scope
+        /// </summary>
+        private class DpiAwarenessScope : IDisposable
+        {
+            private bool dpiAwarenessScopeIsSet;
+            private DPI_HOSTING_BEHAVIOR _originalDpiHostingBehavior = DPI_HOSTING_BEHAVIOR.DPI_HOSTING_BEHAVIOR_INVALID;
+            private readonly DPI_AWARENESS_CONTEXT _originalDpiAwarenessContext;
+
+            /// <summary>
+            ///  Sets <see cref="DPI_AWARENESS_CONTEXT"/> for the thread.
+            /// </summary>
+            /// <param name="context"><see cref="DPI_AWARENESS_CONTEXT"/> to be set on the thread.</param>
+            /// <param name="behavior"><see cref="DPI_HOSTING_BEHAVIOR"/> to be set on the thread.</param>
+            public DpiAwarenessScope(DPI_AWARENESS_CONTEXT context, DPI_HOSTING_BEHAVIOR behavior)
+            {
+                // Full support for DPI_AWARENESS_CONTEXT and mixed mode DPI_HOSTING_BEHAVIOR on the thread is only available after the RS4 OS release.
+                if (!OsVersion.IsWindows10_18030rGreater())
+                {
+                    throw new PlatformNotSupportedException();
```

I think it's safer to fail silently because the UI might or might not be usable, but that's better than a crash. This situation is unlikely to be discovered at development time, so failing fast might not be helpful to developer. But add a `Debug.Assert`

---

Agree. Given I am not certain if DpiScope that we used since. NET Framework 4.7 may not be guarded against OS version check.